### PR TITLE
refactor: validate constraints eagerly

### PR DESCRIPTION
**Pull request by tisonkun**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

TRIVIAL AS IS.

Previously, we can check time index constraint for CREATE EXTERNAL TABLE on parsing (but we do on execution).

Now we check constraints once the condition value are finalized.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

**Comment by MichaelScofield** [comment]
Can you add a case for https://github.com/GreptimeTeam/greptimedb/issues/3471 in sqlness test?

**Comment by tisonkun** [comment]
@MichaelScofield This doesn't fix #3471. Just a front code refactor.

Do you suggest we anyway add a case to indicate the current LTA manner described in #3471?

**Comment by MichaelScofield** [comment]
What's the meaning of "LTA"? 

I mean to add a case in sqlness test that if inputting

```sql
CREATE TABLE IF NOT EXIST t(`Hits` STRING NULL, ts TIMESTAMP(3) NOT NULL, TIME INDEX ("ts"));
```

, the db returns msg indicating it's the "IF NOT EXIST" syntax error instead of "ERROR 1815 (HY000): Missing time index constraint".

File: src/sql/src/parsers/create_parser.rs [hunk 1]
```diff
@@ -62,17 +62,21 @@ impl<'a> ParserContext<'a> {
...
         self.parser
             .expect_keyword(Keyword::TABLE)
-            .context(error::SyntaxSnafu)?;
+            .context(SyntaxSnafu)?;
         let if_not_exists =
             self.parser
                 .parse_keywords(&[Keyword::IF, Keyword::NOT, Keyword::EXISTS]);
         let table_name = self.intern_parse_table_name()?;
         let (columns, constraints) = self.parse_columns()?;
+        if !columns.is_empty() {
```
Comment (waynexia) [inline src/sql/src/parsers/create_parser.rs:71]: It looks like we should check this unconditionally. Empty table is not allowed (and btw is not covered in tests)

Comment (tisonkun) [reply src/sql/src/parsers/create_parser.rs:71]: Interesting empty table is supported for external table:

```sql
CREATE EXTERNAL TABLE IF NOT EXISTS city WITH (location='/var/data/city.csv',format='csv');
```

.. and we have some tests cover this.

Comment (tisonkun) [reply src/sql/src/parsers/create_parser.rs:71]: See `expr_factory::create_external_expr` for an execution time validation.

Comment (waynexia) [reply src/sql/src/parsers/create_parser.rs:71]: That's a bit different. Constraints are more released on external files. Like a `TIME INDEX` is not required.

Comment (tisonkun) [reply src/sql/src/parsers/create_parser.rs:71]: This condition is added later. Previously I check the constraints unconditionally, and it cause the valid SQL above to fail.

Note that this line is for parsing create_external_table statement.

**Comment by MichaelScofield** [comment]
After this PR, what's the error msg if user inputs "create table foo if not exist ..."?

**Comment by tisonkun** [comment]
> After this PR, what's the error msg if user inputs "create table foo if not exist ..."?

```sql
public=> create table if not exists t(ts timestamp time index);
OK 0
public=> create table if not exist t(ts timestamp time index);
ERROR:  Missing time index constraint
public=> 
```

**Review by MichaelScofield — APPROVED** [review]

**Comment by waynexia** [comment]
> > After this PR, what's the error msg if user inputs "create table foo if not exist ..."?
> 
> ```sql
> public=> create table if not exists t(ts timestamp time index);
> OK 0
> public=> create table if not exist t(ts timestamp time index);
> ERROR:  Missing time index constraint
> public=> 
> ```

This error message looks more confusing... Can we avoid it?

**Comment by tisonkun** [comment]
> This error message looks more confusing... Can we avoid it?

This is what #3471 tracked and what I'd try to resolve later. Please read the background. I have explained this purpose many times here 🤣

**Review by waynexia — APPROVED** [review]

**Pull request merged**